File: Data Structures & Algorithms/word-break/submission-6.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] memo = new boolean[n + 1];
        memo[n] = true;

        for (int i = n - 1; i >= 0; i--)
        {
            for (String w : wordDict)
            {
                if (i + w.length() <= n && s.startsWith(w, i))
                {
                    memo[i] = memo[i + w.length()];
                }
                if (memo[i])
                {
                    break;
                }
            }
        }

        return memo[0];
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-6.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        memo = [[None] * 2 for _ in range(len(prices))]

        def dfs(i, buying):
            if i >= len(prices):
                return 0
            if memo[i][buying]:
                return memo[i][buying]
            
            cooldown = dfs(i + 1, buying)
            if buying:
                buy = dfs(i + 1, not buying) - prices[i]
                memo[i][buying] = max(buy, cooldown)
            else:
                sell = dfs(i + 2, not buying) + prices[i]
                memo[i][buying] = max(sell, cooldown)

            return memo[i][buying]
            
        return dfs(0, True)